### PR TITLE
Resolves #15 - Create a function that validate payment response

### LINKED ISSUE #15: Create a helper function to validate webhook payment responses
# Description
Once a user has processed their payments, SISP will return a response about the payment process using our `webhook` as they request implementation in the official [documentation]( https://www.vinti4.cv/documentation.aspx?id=682).

- Create a function that should receive this response and validate if the response was processed successfully or not.
```js
/**
 * VALIDATE PAYMENT RESPONSE
 * @param {Object} responseBody - Required - The response body returned in the webhook by SISP.
 * 
 * @returns {Object} error - An error object if there is an error, otherwise `undefined`. 
 * @returns {Number} error.code - The error code.
 * @returns {String} error.description - Error debug description.
 */
validatePayment(responseBody)
```

- Create reusable `PaymentError` structure with `code` and `description` properties.

| Code: `string`  | Description: `string`               |
|-------|-----------------------------------------------|
| 001   | Payment processing error: Invalid fingerprint |
| 002   | Payment processing error: Cancelled by user   |
| 003   | Payment processing error: Processing error    |

# SISP Payment Validation Implementation Info

1. Success Message Type
If the payment was successfully processed, SISP will send us a success message type that we can access in responseBody.message Type, and this message can be one of the following types:
```js
    // SUCCESS RESPONSE CONSTANTS
    const successMessageTypes = ["8", "10", "M", "P"];
    if (successMessageTypes.includes(responseBody.messageType)) {
       // Validate fingerprint of the result
    } else {
        return ({ code: 003, message: 'Payment processing error: Processing error' });
    }
```

2. Success Message Type
Validate fingerprint of the result. </br> In addition to checking the `messageType` field, it is necessary to check if the FingerPrint received in the resultFingerPrint field is the same as expected, which is calculated according to the use of the function `GenerateFingerPrintResponseSuccessful` see the official [documentation]( https://www.vinti4.cv/documentation.aspx?id=682).
```js
    /*VALIDATE SUCCESS FINGERPRINT*/
    if(responseBody.resultFingerPrint === calculatedFingerPrint)
        // Handle Successful Payment
        return ({ code: 200, message: 'Payment processing successfully' });
    else
        // Handle Unsuccessful Payment: Invalid Response Finger Print	
        return ({ code: 001, message: 'Payment processing error: Invalid fingerprint' });
```

3. Unsuccessful payment
We can get unsuccessful payment if success message type doesn't include any of the type that was described in the point 1, or if the result Fingerprint is different from the calculated fingerprint as showed in point 2 and other error we can get is if the users cancel their payments as the example below:

```js
if(responseBody.UserCancelled === 'true') {
    // User canceled purchase requisition
    return ({ code: 002, message: 'Payment processing error: Cancelled by user' });
}
```

# Testing Requirements
- Ensure all criteria outlined in this ticket is tested.

# README
## Validate Payment Processing Status (convenience method)
Check whether or not a payment was processed successfully.
**Note:** This method is provided as a convenience. See the SISP documentation for the request structure.

### Usage
Pass the Webhook request body to `validatePayment`. Returns an error object containing `code` and `description` if there is an error, otherwise returns `undefined`.

```js
validatePayment(responseBody)
```

#### Possible Errors
| Code: `string`  | Description: `string`               |
|-------|-----------------------------------------------|
| 001   | Payment processing error: Invalid fingerprint |
| 002   | Payment processing error: Cancelled by user   |
| 003   | Payment processing error: Processing error    |

### Example

```js
 app.post("/webhook-handler", (request, response) => {
     const error = sisp.validatePayment(request.body);

     if(error === undefined) {
         // Payment processed successfully
     } 
     else {
         // Payment processing error
         console.log(error.code);
         console.log(error.description);
     }
 }
```

### PR DESCRIPTION
**Pull request by FredSoares**

### CODE REVIEW

File: src/sisp.js [hunk 1]
```diff
@@ -103,6 +103,105 @@ class Sisp {
...
+   * @returns {Object} response.data - This should be the same data returned by SISP, but it will only be returned if the payment is successfully processed, otherwise it will be undefined.
+   */
+  paymentProcessedSuccessfully = (responseBody) => {
+    // SUCCESS RESPONSE CONSTANTS
+    const successMessageTypes = ["8", "10", "M", "P"];
+
+    if (successMessageTypes.includes(responseBody.messageType)) {
+      const posAutCode = this.posAutCode;
+      // Validate fingerprint of the result
+      const calculatedFingerPrint = this.#generateResponseFingerprint(
```
Comment (FabyoSK) [inline src/sisp.js:167]: why you named the function `#generateResponseFingerprint` ?
there a reason to this?

Comment (FredSoares) [reply src/sisp.js:167]: @FabyoSK This is a private method in JS

Comment (FabyoSK) [reply src/sisp.js:167]: Oh, I understood

**Review by FabyoSK — DISMISSED** [review]

File: src/sisp.test.js [hunk 2]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
       const expected = successData.replace(/\s/g, '');
 
       expect(result.replace(/\s/g, '')).toEqual(expected);
     });
   });
+
+  describe('Processing Payment Response', () => {
+    describe('Successful Payment', () => {
+
+      it('should return the correct response if messageType is 8', async () => {
```
Comment (savylopes) [inline src/sisp.test.js]: @FredSoares What should return the correct response?  I think that description can be more descriptive. Please check the repeated scenarios

File: src/sisp.test.js [hunk 3]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
+        const sisp = new Sisp({ posID, posAutCode });
+        const requestFormResult = await sisp.generatePaymentRequestForm(referenceId, total, webhookUrl);
+
+        const formExpected = successData.replace(/\s/g, '');
+
+        const result = await sisp.paymentProcessedSuccessfully(paymentSuccessData);
+        const expected = { code: 200, message: 'Payment processing success', data: paymentSuccessData };
+
+        expect(requestFormResult.replace(/\s/g, '')).toEqual(formExpected);
+        expect(result).toEqual(expected);
```
Comment (savylopes) [inline src/sisp.test.js:113]: Maybe is better we add something like this to be more aligned with  the description: `expect(result.data.messageType).toEqual(expected.data.messageType);` then if this fail we will se the message for example: this expected 8 but got 50. What do you think @FredSoares. Please check the repeated scenarios

File: src/sisp.test.js [hunk 4]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
```
Comment (savylopes) [inline src/sisp.test.js:151]: Maybe is better we add something like this to be more aligned with the description: expect(result.data.messageType).toEqual(expected.data.messageType); then if this fail we will se the message for example: this expected 10 but got 50. What do you think @FredSoares

File: src/sisp.test.js [hunk 5]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
```
Comment (savylopes) [inline src/sisp.test.js:191]: Maybe is better we add something like this to be more aligned with the description: expect(result.data.messageType).toEqual(expected.data.messageType); then if this fail we will se the message for example: this expected M but got 50. What do you think @FredSoares

File: src/sisp.test.js [hunk 6]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
```
Comment (savylopes) [inline src/sisp.test.js:231]: Maybe is better we add something like this to be more aligned with the description: expect(result.data.messageType).toEqual(expected.data.messageType); then if this fail we will se the message for example: this expected P but got 50. What do you think @FredSoares

File: src/sisp.test.js [hunk 7]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
+        const sisp = new Sisp({ posID, posAutCode });
+        const requestFormResult = await sisp.generatePaymentRequestForm(referenceId, total, webhookUrl);
+
+        const formExpected = successData.replace(/\s/g, '');
+
+        const result = sisp.paymentProcessedSuccessfully(paymentSuccessData);
+        const expected = { code: 400, message: 'Payment processing error' };
+
+        expect(requestFormResult.replace(/\s/g, '')).toEqual(formExpected);
+        expect(result).toEqual(expected);
```
Comment (savylopes) [inline src/sisp.test.js:278]: Maybe is better we add something like this to be more aligned with the description: expect(result.data.messageType).toEqual(expected.data.messageType) for each success message; then if this fail we will se the message for example: this expected 8 but got 50. What do you think @FredSoares

File: src/sisp.test.js [hunk 8]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
+        const formExpected = successData.replace(/\s/g, '');
+
+        const result = await sisp.paymentProcessedSuccessfully(paymentSuccessData);
+        const expected = { code: 200, message: 'Payment processing success', data: paymentSuccessData };
+
+        expect(requestFormResult.replace(/\s/g, '')).toEqual(formExpected);
+        expect(result).toEqual(expected);
+      });
+
+      it('should return the correct response if messageType is 10', async () => {
```
Comment (savylopes) [inline src/sisp.test.js:114]: @FredSoares What should return the correct response? I think that description can be more descriptive.

File: src/sisp.test.js [hunk 9]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
+        const formExpected = successData.replace(/\s/g, '');
+
+        const result = await sisp.paymentProcessedSuccessfully(paymentSuccessData);
+        const expected = { code: 200, message: 'Payment processing success', data: paymentSuccessData };
+
+        expect(requestFormResult.replace(/\s/g, '')).toEqual(formExpected);
+        expect(result).toEqual(expected);
+      });
+
+      it('should return the correct response if messageType is M', async () => {
```
Comment (savylopes) [inline src/sisp.test.js:154]: @FredSoares What should return the correct response? I think that description can be more descriptive.

File: src/sisp.test.js [hunk 10]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
+        const formExpected = successData.replace(/\s/g, '');
+
+        const result = await sisp.paymentProcessedSuccessfully(paymentSuccessData);
+        const expected = { code: 200, message: 'Payment processing success', data: paymentSuccessData };
+
+        expect(requestFormResult.replace(/\s/g, '')).toEqual(formExpected);
+        expect(result).toEqual(expected);
+      });
+
+      it('should return the correct response if messageType is P', async () => {
```
Comment (savylopes) [inline src/sisp.test.js:194]: @FredSoares What should return the correct response? I think that description can be more descriptive.

File: src/sisp.test.js [hunk 11]
```diff
@@ -59,12 +59,218 @@ describe('SISP Payment', () => {
...
+
+        const result = await sisp.paymentProcessedSuccessfully(paymentSuccessData);
+        const expected = { code: 200, message: 'Payment processing success', data: paymentSuccessData };
+
+        expect(requestFormResult.replace(/\s/g, '')).toEqual(formExpected);
+        expect(result).toEqual(expected);
+      });
+    });
+    describe('Error Payment', () => {
+      it('should fail if messageType is different from "8", "10", "M" or "P"', async () => {
```
Comment (savylopes) [inline src/sisp.test.js:235]: @FredSoares What should return the correct response? I think that description can be more descriptive.

**Review by savylopes — CHANGES_REQUESTED** [review]

File: src/sisp.test.js (hunk 2 above)
Comment (FredSoares) [reply src/sisp.test.js]: @savylopes As you can see in the description of this issue, we just have one success case and the response will be always the same, and if you look at this test you will see the response will be the expected case `const expected = { code: 200, message: 'Payment processing success', data: paymentSuccessData }; `

File: src/sisp.test.js (hunk 3 above)
Comment (FredSoares) [reply src/sisp.test.js:113]: @savylopes If this fail `Jest` will show the expected response and the received response with their differences.

File: src/sisp.test.js (hunk 7 above)
Comment (FredSoares) [reply src/sisp.test.js:278]: @savylopes If this fail `Jest` will show the expected response and the received response with their differences

File: src/sisp.test.js (hunk 2 above)
Comment (savylopes) [reply src/sisp.test.js]: It is just for description if it fail the person that run the test can know exactly what return the wrong response if you improve it

File: src/sisp.test.js (hunk 3 above)
Comment (savylopes) [reply src/sisp.test.js:113]: Yes you right. thank you

File: src/sisp.test.js (hunk 7 above)
Comment (savylopes) [reply src/sisp.test.js:278]: Yes. You right

**Review by savylopes — CHANGES_REQUESTED** [review]

**Review by savylopes — DISMISSED** [review]

**Review by FabyoSK — DISMISSED** [review]

File: src/sisp.js [hunk 12]
```diff
@@ -103,6 +103,105 @@ class Sisp {
...
+        responseBody.merchantRespTimeStamp,
+        responseBody.merchantRespReferenceNumber,
+        responseBody.merchantRespEntityCode,
+        responseBody.merchantRespClientReceipt,
+        responseBody.merchantRespAdditionalErrorMessage.trim(),
+        responseBody.merchantRespReloadCode
+      );
+
+      // Validade success fingerprint
+      if (responseBody.resultFingerPrint === calculatedFingerPrint) {
```
Comment (lcsfort) [inline src/sisp.js:187]: @FredSoares according to your ticket description I don't see the scenario where `3 - Unsuccessful payment`, the user canceled purchase requisition

Comment (FredSoares) [reply src/sisp.js:187]: @luis737 Good catch thanks

**Review by FabyoSK — DISMISSED** [review]

**Review by lcsfort — DISMISSED** [review]

**Review by savylopes — DISMISSED** [review]

**Review by FabyoSK — DISMISSED** [review]

**Review by lcsfort — DISMISSED** [review]

**Review by savylopes — DISMISSED** [review]

**Review by FabyoSK — APPROVED** [review]

**Review by lcsfort — APPROVED** [review]

**Review by savylopes — APPROVED** [review]

**Review by trismegis7us — APPROVED** [review]

**Pull request merged**